core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveBroadcastToPartitionedJoin.java

```diff
@@ -0,0 +1,460 @@
+    {
+        return node.getScope() == REMOTE
+                && node.getType() == REPARTITION
+                && node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION)
+                && node.getSources().size() == 1;
+    }
+
+    private static boolean isBroadcastRemoteExchange(ExchangeNode node)
+    {
+        return node.getScope() == REMOTE
+                && node.getType() == REPARTITION
+                && node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION)
+                && node.getSources().size() == 1;
+    }
+
+    public AdaptiveBroadcastToPartitionedJoin(CostComparator costComparator, TaskCountEstimator taskCountEstimator)
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        // This rule is only enabled in case of FTE
+        return getRetryPolicy(session) == TASK
+                && isFaultTolerantExecutionAdaptiveBroadcastToPartitionedJoinEnabled(session);
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        if (mustReplicate(node, context)) {
+            return Result.empty();
+        }
+        boolean isExtraRemoteExchangeNeededAtProbeSide = captures.getOptional(LEFT_EXCHANGE_NODE).isEmpty();
+        boolean isExtraRemoteExchangeNeededAtBuildSide = captures.getOptional(RIGHT_EXCHANGE_NODE).isEmpty();
+
+        boolean canChangeDistribution = changeDistributionToPartitioned(
+                node,
+                context,
+                isExtraRemoteExchangeNeededAtProbeSide,
+                isExtraRemoteExchangeNeededAtBuildSide);
+
+        if (!canChangeDistribution) {
+            return Result.empty();
+        }
+
+        return Result.ofPlanNode(convertToPartitionedJoin(node, context));
+    }
+
+    private static boolean mustReplicate(JoinNode joinNode, Context context)
+    {
+        JoinType type = joinNode.getType();
+        if (joinNode.getCriteria().isEmpty() && (type == INNER || type == LEFT)) {
+            // There is nothing to partition on
+            return true;
+        }
+        return isAtMostScalar(joinNode.getRight(), context.getLookup());
+    }
+
+    private boolean changeDistributionToPartitioned(
+            JoinNode joinNode,
+            Context context,
+            boolean isExtraRemoteExchangeNeededAtProbeSide,
+            boolean isExtraRemoteExchangeNeededAtBuildSide)
+    {
+        DataSize joinMaxBroadcastTableSize = getJoinMaxBroadcastTableSize(context.getSession());
+
+        PlanNodeWithCost replicatedJoinCost = getJoinNodeWithCost(
```

Comment:
reming me where are we taking into account that some of the progress has been made alrady with current plan shape, and if we replan we need to start from scratch.